rust/src/tests/errors.rs

```diff
@@ -0,0 +1,589 @@
+    where
+        F: Fn(&Path) -> nix::Result<T>,
+    {
+        assert_eq!(f(p1).unwrap_err(), Errno::ELOOP);
+        assert_eq!(f(p2).unwrap_err(), Errno::ELOOP);
+    }
+    fn assert_eloop_final<F, T: Debug>(p1: &Path, p2: &Path, f: F)
+    where
+        F: Fn(&Path) -> nix::Result<T>,
+    {
+        assert_eq!(f(&p1.join("test")).unwrap_err(), Errno::ELOOP);
+        assert_eq!(f(&p2.join("test")).unwrap_err(), Errno::ELOOP);
+    }
+    fn assert_eloop_link<F, T: Debug>(p1: &Path, p2: &Path, p3: &Path, f: F)
+    where
+        F: Fn(&Path, &Path) -> nix::Result<T>,
+    {
+        assert_eq!(f(&p1.join("test"), &p3).unwrap_err(), Errno::ELOOP);
+        assert_eq!(f(&p2.join("test"), &p3).unwrap_err(), Errno::ELOOP);
+        assert_eq!(f(&p3, &p1.join("test")).unwrap_err(), Errno::ELOOP);
+        assert_eq!(f(&p3, &p2.join("test")).unwrap_err(), Errno::ELOOP);
+    }
+    fn assert_eloop_all<F, T: Debug>(p1: &Path, p2: &Path, f: F)
+    where
+        F: Fn(&Path) -> nix::Result<T>,
+    {
+        assert_eloop_folder(p1, p2, &f);
+        assert_eloop_final(p1, p2, f);
+    }
+
+    // TODO: Add rmdir upstream
+    // assert_eloop(&p1, &p2, |p| rmdir(p, Mode::empty()));
+    assert_eloop_final(&p1, &p2, |p| chflags(p, FileFlag::empty()));
+    assert_eloop_all(&p1, &p2, |p| chmod(p, Mode::empty()));
+    assert_eloop_final(&p1, &p2, |p| lchmod(p, Mode::empty()));
+    assert_eloop_all(&p1, &p2, |p| {
+        chown(
+            p,
+            Some(User::from_name("nobody").unwrap().unwrap().uid),
+            None,
+        )
+    });
+    assert_eloop_final(&p1, &p2, |p| {
+        lchown(
+            p,
+            Some(User::from_name("nobody").unwrap().unwrap().uid),
+            None,
+        )
+    });
+    assert_eloop_link(&p1, &p2, &p3, |p1, p2| link(p1, p2));
+    assert_eloop_final(&p1, &p2, |p| mkdir(p, Mode::empty()));
+    assert_eloop_final(&p1, &p2, |p| mkfifo(p, Mode::empty()));
+    assert_eloop_final(&p1, &p2, |p| mknod(p, SFlag::S_IFCHR, Mode::empty(), 0));
+    assert_eloop_final(&p1, &p2, |p| open(p, OFlag::empty(), Mode::empty()));
+    assert_eloop_link(&p1, &p2, &p3, |p1, p2| rename(p1, p2));
+    assert_eloop_final(&p1, &p2, |p| symlink(Path::new("test"), p));
+    assert_eloop_final(&p1, &p2, |p| truncate(p, 0));
+    assert_eloop_final(&p1, &p2, |p| unlink(p));
+}
+
+crate::test_case! {enotdir => [Regular, Fifo, Block, Char, Socket]}
+fn enotdir(ctx: &mut TestContext, ft: FileType) {
+    let base_path = ctx.create(ft).unwrap();
+    let path = base_path.join("previous_not_dir");
+    let dir = ctx.create(FileType::Dir).unwrap();
+
+    fn assert_enotdir<T: Debug, F>(path: &Path, f: F)
+    where
+        F: Fn(&Path) -> nix::Result<T>,
+    {
+        assert_eq!(f(path).unwrap_err(), Errno::ENOTDIR);
+    }
+
+    // TODO: Add rmdir upstream
+    // assert_enotdir(&path, |p| rmdir(p, Mode::empty()));
+    assert_enotdir(&path, |p| chflags(p, FileFlag::empty()));
```

Comment:
The assert_enotdir function doesn't buy much.  I think it's more readable just to inline it:
```rust
assert_eq!(chmod(p, Mode::empty()), Error::ENOTDIR);
```